packages/aws-cdk/lib/exec.ts

```diff
@@ -0,0 +1,154 @@
+ */
+function versionCheckResponse(response: cxapi.SynthesizeResponse): cxapi.SynthesizeResponse {
+  if (!response.version) {
+    // tslint:disable-next-line:max-line-length
+    throw new Error(`CDK Framework >= ${cxapi.PROTO_RESPONSE_VERSION} is required in order to interact with this version of the Toolkit.`);
+  }
+
+  const frameworkVersion = semver.coerce(response.version);
+  const toolkitVersion = semver.coerce(cxapi.PROTO_RESPONSE_VERSION);
+
+  // Should not happen, but I don't trust this library 100% either, so let's check for it to be safe
+  if (!frameworkVersion || !toolkitVersion) { throw new Error('SemVer library could not parse versions'); }
+
+  if (semver.gt(frameworkVersion, toolkitVersion)) {
+    throw new Error(`CDK Toolkit >= ${response.version} is required in order to interact with this program.`);
+  }
+
+  if (semver.lt(frameworkVersion, toolkitVersion)) {
+    // Toolkit protocol is newer than the framework version, and we KNOW the
+    // version. This is a scenario in which we could potentially do some
+    // upgrading of the response in the future.
+    //
+    // For now though, we simply reject old responses.
+    throw new Error(`CDK Framework >= ${cxapi.PROTO_RESPONSE_VERSION} is required in order to interact with this version of the Toolkit.`);
+  }
+
+  return response;
+}
+
+/**
+ * If we don't have region/account defined in context, we fall back to the default SDK behavior
+ * where region is retreived from ~/.aws/config and account is based on default credentials provider
+ * chain and then STS is queried.
+ *
+ * This is done opportunistically: for example, if we can't acccess STS for some reason or the region
+ * is not configured, the context value will be 'null' and there could failures down the line. In
+ * some cases, synthesis does not require region/account information at all, so that might be perfectly
+ * fine in certain scenarios.
+ *
+ * @param context The context key/value bash.
+ */
+async function populateDefaultEnvironmentIfNeeded(aws: SDK, context: any) {
+  if (!(cxapi.DEFAULT_REGION_CONTEXT_KEY in context)) {
+    context[cxapi.DEFAULT_REGION_CONTEXT_KEY] = await aws.defaultRegion();
+    debug(`Setting "${cxapi.DEFAULT_REGION_CONTEXT_KEY}" context to`, context[cxapi.DEFAULT_REGION_CONTEXT_KEY]);
+  }
+
+  if (!(cxapi.DEFAULT_ACCOUNT_CONTEXT_KEY in context)) {
+    context[cxapi.DEFAULT_ACCOUNT_CONTEXT_KEY] = await aws.defaultAccount();
+    debug(`Setting "${cxapi.DEFAULT_ACCOUNT_CONTEXT_KEY}" context to`, context[cxapi.DEFAULT_ACCOUNT_CONTEXT_KEY]);
+  }
+}
+
+/**
+ * Make sure the 'app' is an array
+ *
+ * If it's a string, split on spaces as a trivial way of tokenizing the command line.
+ */
+function appToArray(app: any) {
+  return typeof app === 'string' ? app.split(' ') : app;
+}
+
+/**
+ * Guess the executable from the command-line argument
+ */
+async function guessExecutable(commandLine: string[]) {
```

Comment:
I was actually thinking we can design this a bit more generally and make people happy. For example, if app is a `.js` file, we can determine that it can be executed with `node`, if it's a `.dll`, we can use `dotnet`, etc. What do you think?

I was also thinking that we can introduce auto-detection (i.e. if there's a `package.json` file, or `pom.xml`, we can auto-detect where the CDK app is), but that's a later thing.

For now, I think adding support for `.js` could be nice  and maybe implement this a bit more generically (extension to programming mapping).

Reply:
Yes, that seems reasonable.